### ISSUE TITLE
Incorrect JDK version activation when running project within Eclipse/m2e

### ISSUE THREAD

**Issue by harbulot** [eclipse-m2e/m2e-core#497]
It seems that the JDK version number used to activate a profile doesn't match the version of the project execution environment / JDK version.

* Eclipse Version: 2021-12 (4.22.0) (running on JDK 17).
* M2E Maven Integration for Eclipse Core 1.19.0.20211116-0804

Sample `pom.xml`:

```xml
<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
    <modelVersion>4.0.0</modelVersion>
    <groupId>test003</groupId>
    <artifactId>test003</artifactId>
    <version>0.0.1-SNAPSHOT</version>
    <build>
        <plugins>
            <plugin>
                <groupId>org.apache.maven.plugins</groupId>
                <artifactId>maven-compiler-plugin</artifactId>
                <version>3.9.0</version>
                <configuration>
                    <source>1.8</source>
                    <target>1.8</target>
                    <testSource>1.8</testSource>
                    <testTarget>1.8</testTarget>
                </configuration>
            </plugin>
        </plugins>
    </build>

    <profiles>
        <profile>
            <id>jdk-15-and-above</id>
            <activation>
                <jdk>[15,)</jdk>
            </activation>
            <dependencies>
                <!-- JavaScript -->
                <dependency>
                    <groupId>org.openjdk.nashorn</groupId>
                    <artifactId>nashorn-core</artifactId>
                    <version>15.3</version>
                </dependency>
            </dependencies>
        </profile>
    </profiles>
</project>
```

Sample class:

```java
package test003;

import javax.script.ScriptEngine;
import javax.script.ScriptEngineManager;
import javax.script.ScriptException;

public class MyClass {
    public MyClass() throws ScriptException {
        ScriptEngine scriptEngine = new ScriptEngineManager().getEngineByName("nashorn");
        Object v = scriptEngine.eval("1;");
        System.out.println("Result: " + v);
    }

    public static void main(String[] args) throws Exception {
        new MyClass();
    }
}
```

Output:

> Exception in thread "main" java.lang.UnsupportedClassVersionError: org/openjdk/nashorn/api/scripting/NashornScriptEngineFactory has been compiled by a more recent version of the Java Runtime (class file version 55.0), this version of the Java Runtime only recognizes class file versions up to 52.0

Essentially, the `jdk-15-and-above` profile, which should only be activated for JDK versions 15 and above (`<jdk>[15,)</jdk>`) has been activated, and this incompatible dependency has been included on the classpath, even though the project is compiled and running with a JDK version 8.

![image](https://user-images.githubusercontent.com/80994/149674680-77aa6060-3313-4a65-8efb-a702ead168f3.png)

![image](https://user-images.githubusercontent.com/80994/149674683-f8058d4c-5c46-4ee1-884a-c923b72d7c8b.png)

**Comment by laeubi** [comment]
This seems related to

- https://github.com/eclipse-m2e/m2e-core/issues/116

because currently m2e uses your default JRE, could you check that this is a real Java-8JDK configured as default VM and mapped to the appropriate Java-EE?

**Comment by mickaelistria** [comment]
I don't think it's much related to #116 , but it is probably related to #248

**Comment by harbulot** [comment]
Yes, it's an actual JDK 8 as far as I know:

![image](https://user-images.githubusercontent.com/80994/149752555-13d87fce-719c-4112-9850-0b87c77e5fc0.png)

![image](https://user-images.githubusercontent.com/80994/149752566-3ddca657-924d-4c1b-a2bc-ed4cadf84061.png)

I think it looks related to #248 and https://bugs.eclipse.org/bugs/show_bug.cgi?id=432992 indeed.

**Comment by laeubi** [comment]
> Yes, it's an actual JDK 8 as far as I know:

This looks correct, just for the sake of testing, can you remove the jdk17 and see if this changes anything?

**Comment by harbulot** [comment]
I've removed the JDK 17 from the Installed JREs, restarted Eclipse, but the `jdk-15-and-above` profile is still activated.

**Comment by laeubi** [comment]
@harbulot thanks for verification, do you think you can propose a patch for this?

**Comment by harbulot** [comment]
Sorry, I don't think I can be in a position to provide a patch since I'm completely new to this code-base. I'm unfortunately not sufficiently familiar with the various package/classes of m2e, Eclipse and Maven.

That said, I've tried to dig a little, so here is what might be relevant.

Since I knew that `<activation><property><name>!m2e.version</name></property></activation>` is a way to activate plugins when `m2e` is used, I've simply searched for `"m2e.version"` in the repository.
This led me to [`org.eclipse.m2e.core.internal.embedder.MavenImpl#createExecutionRequest()`](https://github.com/eclipse-m2e/m2e-core/blob/11a45fb6a832a6fcf38f0410ebc553e9c466d1ff/org.eclipse.m2e.core/src/org/eclipse/m2e/core/internal/embedder/MavenImpl.java#L252).

Line 252 is of interest:

```java
MavenExecutionContext.populateSystemProperties(request);
```

Following up on that, we can see in [`MavenExecutionContext`](https://github.com/eclipse-m2e/m2e-core/blob/11a45fb6a832a6fcf38f0410ebc553e9c466d1ff/org.eclipse.m2e.core/src/org/eclipse/m2e/core/internal/embedder/MavenExecutionContext.java#L229):

```java
  public static void populateSystemProperties(MavenExecutionRequest request) {
    // temporary solution for https://issues.sonatype.org/browse/MNGECLIPSE-1607
    // oddly, there are no unit tests that fail if this is commented out
    if(request.getSystemProperties() == null || request.getSystemProperties().isEmpty()) {
      Properties systemProperties = new Properties();
      EnvironmentUtils.addEnvVars(systemProperties);
      copyProperties(systemProperties, System.getProperties());
      MavenProperties.setProperties(systemProperties);
      request.setSystemProperties(systemProperties);
    }
  }
```

Here seems to lie the core of the problem: `copyProperties(systemProperties, System.getProperties())` copies the current `System.getProperties()` (which almost certainly come from the JVM running Eclipse itself).


`MavenImpl.createExecutionRequest()` is effectively called from [`MavenExecutionContext.execute(MavenProject, ...)`](https://github.com/eclipse-m2e/m2e-core/blob/11a45fb6a832a6fcf38f0410ebc553e9c466d1ff/org.eclipse.m2e.core/src/org/eclipse/m2e/core/internal/embedder/MavenExecutionContext.java#L107) (indirectly, via `newExecutionRequest()`).

Note that `populateSystemProperties` seems to be called again anyway in this method (after `MavenExecutionContext.newExecutionRequest()` / `MavenImpl.createExecutionRequest()`):

```java
      if(request == null) {
        request = newExecutionRequest();
      }
      maven.populateDefaults(request);
      populateSystemProperties(request);
```

This is just a guess, but I think some of those properties should be overwritten in the request, using configuration coming from the Project (or even the Run Configuration if that's possible). Some of the properties that come to mind:

* `java.version`
* `java.home`
* `java.vm.specification.version`
* `java.specification.version`
* `java.class.version`
* `java.runtime.version`

Those should in principle come from the project's Execution Environment JRE, not the System Properties of the running Eclipse. (I can also see a situation where one could want to pick up the Run Configuration Execution Environment, which may differ from the Project's Execution Environment, but that could be a step too far.)

Unfortunately, I must admit I don't know enough about the Eclipse internals to know where those values could be obtained from.

**Comment by laeubi** [comment]
After thinking a bit more about this, I think this can not work as expected. m2e (and therefore the mojo/incremental build) is running inside eclipse so "faking" properties that claim a different vm won't make a real difference and probably even harm more than help.

What would help is, that m2e implements the ProfileManger and determines profile activation that take java version into account by the java version set for the project.

**Comment by HannesWell** [comment]
> After thinking a bit more about this, I think this can not work as expected. m2e (and therefore the mojo/incremental build) is running inside eclipse so "faking" properties that claim a different vm won't make a real difference and probably even harm more than help.
> 
> What would help is, that m2e implements the ProfileManger and determines profile activation that take java version into account by the java version set for the project.


That is indeed a systematic problem of the way how M2E works. So yes injecting modified properties is indeed dangerous. But wouldn't it be dangerous as well to have a profile activated based on the JVM-version that would not have been activated with the actual JVM? Because the plug-ins/executions configured there could make assumptions about the mentioned variables or the capabilities of the VM.
For example if you turn around the example from the initial comment and have a profile activated for java<15 were one expects that nashorn is present and does something with java-script. This will then blow up as well.
In most cases it isn't a problem to have a more recent JRE/JDK but in cases like that it is.

**Comment by laeubi** [comment]
This is just another incarnation of runtime vs compile JVM... So the problem is that a wrong dependency is selected compared to the JVM that is used to run code (here a main method). So it would help for such cases but still not work for others unless we want to span a "maven-server" for each project (group) with a dedicated JVM... still this won't cover the case where one launches "something" like a main from other toolings.

**Comment by kwin** [comment]
> so "faking" properties that claim a different vm won't make a real difference and probably even harm more than help.

I disagree with that. Using the same java.home value when running via Launch Maven and incremental build/project configuration inside Eclipse helps to solve most of the issues (including this one here)

**Comment by kwin** [comment]
> Just assume the case you are using JDK17 to run Eclipse and have a project that uses Java19 to compile and run the build ...

I agree, there should be big WARN in case the Maven Execution JRE (derived from m-enforcer-p, requireJavaVersion) is higher than Java17. This is an edge case, as most Maven projects I know of are supporting building with Java 17 or older, though. This is already a problem right now, even if the pom.xml does not refer to any `java.home` variable. 

> java-home is just a very poor reference to use in a maven build

There are some edge cases which unfortunately require that, e.g. adding `tools.jar` to the compile/runtime classpath (like outlined in https://stackoverflow.com/a/3080695).

**Comment by laeubi** [comment]
> > Just assume the case you are using JDK17 to run Eclipse and have a project that uses Java19 to compile and run the build ...
> 
> I agree, there should be big WARN in case the Maven Execution JRE (derived from m-enforcer-p, requireJavaVersion) is higher than Java17. This is an edge case, as most Maven projects I know of are supporting building with Java 17 or older, though. This is already a problem right now, even if the pom.xml does not refer to any `java.home` variable.

That's why I said, I can imagine that it works for resolving dependencies conditionally (and that's what reported here), but one can simply not assume it works in embedded mode to "lie" about the used JVM, it is always the one Eclipse is started with, and for

> There are some edge cases which unfortunately require that, e.g. adding `tools.jar` to the compile/runtime classpath (like outlined in https://stackoverflow.com/a/3080695).

such edge-case it might then be more appropriate to start eclipse with this special JVM (or more specific jdk) than to combine an arbitrary `tools.jar`, I suspect that no jvm vendor will ever claim that the `tools.jar` is compatible to different vendors or versions.... and it seems that even java itself has deprecated that approach now for a long time.

**Comment by kwin** [comment]
Tools.jar has a common API used by all vendors. Still for licensing reasons there is no artifact for it in Maven Central, therefore this needs to be referenced like this. This is totally sufficient for performing incremental builds of classes referencing that API.

**Comment by laeubi** [comment]
If it is "common to all vendors" why does it then matter what tools.jar is actually used? Anyways this issue is not about `tools.jar` anyways...

**Comment by kwin** [comment]
It doesn’t matter, but as I said tools.jar is only available with JDK 8 or below. So this can never work with the JDK/JRE used with Eclipse
